pkg/retriever/graphsyncretriever.go

```diff
@@ -25,43 +28,55 @@ type CandidateCallback func(types.RetrievalCandidate) error
 type CandidateErrorCallback func(types.RetrievalCandidate, error)
 
 type GetStorageProviderTimeout func(peer peer.ID) time.Duration
-type IsAcceptableQueryResponse func(peer peer.ID, req types.RetrievalRequest, queryResponse *retrievaltypes.QueryResponse) bool
 
-type queryCandidate struct {
-	*retrievaltypes.QueryResponse
+type connectCandidate struct {
+	types.RetrievalCandidate
 	Duration time.Duration
 }
 
 // queryCompare compares two QueryResponses and returns true if the first is
 // preferable to the second. This is used for the PriorityWaitQueue that will
 // prioritise execution of retrievals if two queries are available to compare
 // at the same time.
-var queryCompare prioritywaitqueue.ComparePriority[*queryCandidate] = func(a, b *queryCandidate) bool {
-	// Always prefer unsealed to sealed, no matter what
-	if a.UnsealPrice.IsZero() && !b.UnsealPrice.IsZero() {
+var candidateCompare prioritywaitqueue.ComparePriority[connectCandidate] = func(a, b connectCandidate) bool {
+	mdA, ok := a.Metadata.Get(multicodec.TransportGraphsyncFilecoinv1).(*metadata.GraphsyncFilecoinV1)
+	if !ok {
+		return false
+	}
+	mdB, ok := b.Metadata.Get(multicodec.TransportGraphsyncFilecoinv1).(*metadata.GraphsyncFilecoinV1)
+	if !ok {
 		return true
 	}
 
-	// Select lower price, or continue if equal
-	aTotalCost := totalCost(a.QueryResponse)
-	bTotalCost := totalCost(b.QueryResponse)
-	if !aTotalCost.Equals(bTotalCost) {
-		return aTotalCost.LessThan(bTotalCost)
+	// prioritize verified deals over not verified deals
+	if mdA.VerifiedDeal != mdB.VerifiedDeal {
+		return mdA.VerifiedDeal
 	}
 
-	// Select smaller size, or continue if equal
-	if a.Size != b.Size {
-		return a.Size < b.Size
+	// prioritize fast retrievel over not fast retrieval
+	if mdA.FastRetrieval && !mdB.FastRetrieval {
+		return true
```

Comment:
I'd love to have data on the impact of this; I hope it actually makes a difference. Great we're finally doing it.